cpp/src/arrow/acero/source_node_test.cc

```diff
@@ -0,0 +1,161 @@
+
+#include <gtest/gtest.h>
+
+#include "arrow/acero/exec_plan.h"
+#include "arrow/acero/map_node.h"
+#include "arrow/acero/options.h"
+#include "arrow/acero/test_nodes.h"
+
+namespace arrow {
+namespace acero {
+
+struct PauseThenStopNodeOptions : public ExecNodeOptions {
+  explicit PauseThenStopNodeOptions(int num_pass) : num_pass(num_pass) {}
+
+  int num_pass;
+};
+
+template <typename ThisNode>
+struct PauseThenStopNode : public MapNode {
+  PauseThenStopNode(ExecPlan* plan, std::vector<ExecNode*> inputs,
+                    std::shared_ptr<Schema> output_schema,
+                    const PauseThenStopNodeOptions& options)
+      : MapNode(plan, inputs, output_schema), num_pass(options.num_pass) {}
+
+  static Result<ExecNode*> Make(ExecPlan* plan, std::vector<ExecNode*> inputs,
+                                const ExecNodeOptions& options) {
+    RETURN_NOT_OK(ValidateExecNodeInputs(plan, inputs, 1, ThisNode::kKindName));
+    auto bp_options = static_cast<const PauseThenStopNodeOptions&>(options);
+    if (bp_options.num_pass < 2) {
+      return Status::Invalid("num_pass must be at least 2");
+    }
+    return plan->EmplaceNode<ThisNode>(plan, inputs, inputs[0]->output_schema(),
+                                       bp_options);
+  }
+
+  const char* kind_name() const override { return ThisNode::kKindName; }
+  Result<ExecBatch> ProcessBatch(ExecBatch batch) override {
+    if (num_pass == 1) {
+      inputs()[0]->PauseProducing(this, 1);
+      ARROW_RETURN_NOT_OK(static_cast<ThisNode*>(this)->DoStopProducing());
+    }
+    if (num_pass > 0) --num_pass;
+    return batch;
+  }
+
+  int num_pass;
+};
+
+struct PauseThenStopInputNode : public PauseThenStopNode<PauseThenStopInputNode> {
+  static constexpr bool kPlanExitsEarly = false;
+
+  static constexpr const char* kKindName = "PauseThenStopInputNode";
+  static constexpr const char* kFactoryName = "pause_then_stop_input";
+
+  static Status Register() {
+    auto exec_reg = default_exec_factory_registry();
+    if (!exec_reg->GetFactory(kFactoryName).ok()) {
+      ARROW_RETURN_NOT_OK(
+          exec_reg->AddFactory(kFactoryName, PauseThenStopInputNode::Make));
+    }
+    return Status::OK();
+  }
+
+  using PauseThenStopNode::PauseThenStopNode;
+
+  Status DoStopProducing() { return inputs()[0]->StopProducing(); }
+};
+
+struct PauseThenStopPlanNode : public PauseThenStopNode<PauseThenStopPlanNode> {
+  static constexpr bool kPlanExitsEarly = true;
+
+  static constexpr const char* kKindName = "PauseThenStopPlanNode";
+  static constexpr const char* kFactoryName = "pause_then_stop_plan";
+
+  static Status Register() {
+    auto exec_reg = default_exec_factory_registry();
+    if (!exec_reg->GetFactory(kFactoryName).ok()) {
+      ARROW_RETURN_NOT_OK(
+          exec_reg->AddFactory(kFactoryName, PauseThenStopPlanNode::Make));
+    }
+    return Status::OK();
+  }
+
+  using PauseThenStopNode::PauseThenStopNode;
+
+  Status DoStopProducing() {
+    plan()->StopProducing();
```

Comment:
A node that shutdown the entire plan looks weird - I don't think any other node does this. Can we trigger the StopProducing the normal way? (i.e., from the thread that invokes the execution).

Reply:
In general, I would prefer the test case to as close to real scenario as possible. I.e., the real scenario is:
(1) Back pressure is applied 
(2) ExecPlan::StopProducing() by user
(3) Plan doesn't stop

So the test case here ideally should be close to that case

Reply:
> A node that shutdown the entire plan looks weird

Wouldn't a fatal error detected by the node be a good reason to do this?

> I don't think any other node does this

Perhaps. But it should still useful to test this -  a call to `StopProducing` is supposed to work from anywhere.

> Can we trigger the StopProducing the normal way? (i.e., from the thread that invokes the execution).

Sure. Why not as an additional test?